index.rst

```diff
@@ -58,24 +58,316 @@ We describe the design and implementation of the LSST Alert Distribution System,
+The latest release of ``alert_stream`` uses Kafka and Zookeeper from
+Confluent platform release 4.1.1.
+The producer used for generating and sending data to Kafka and
+template scripts for consumers of the stream are provided in the GitHub
+repository at https://github.com/lsst-dm/alert_stream,
+which can also be built as a Docker image and deployed as containers.
+`DMTN-028 <https://dmtn-028.lsst.io>`_ provides details about benchmarking
+deployment of the different components.
 
 Alert Filtering
 ================
 
+Alert filters can be written using simple Python functions
+acting solely on the contents of the alert packets, as
+described in the Data Products Definition Document.
+Using Python allows scientific users to write complex
+functions that are more powerful and simpler to write than
+equivalent filters in alternatives such as SQL.
+If permitted, scientists could also potentially use modules outside of the
+standard library to deploy, e.g., machine learning algorithms on alerts.
+User-provided code, however, presents challenges.
+The filter submission process needs to be controlled.
+Access to the host system needs to be restricted to prevent malicious
+activity and security concerns.
+To mitigate the concerns of user-provided code, in this design
+each filter runs separately and isolated in its own container.
+The filter is only allowed access to the approved libraries available in
+the Python environment provided in the container, and permissions
+of the container can be restricted so as to minimize effects on the host.
+
+A filter is constructed as a consumer of the Kafka topic containing the
+full stream and a producer back to a Kafka topic for the filtered stream.
+This filtered stream can then be read by another consumer or directed
+to another output method for storage.
+The downstream consumer of the filtered stream can itself be another
+filter, meaning that this scheme allows simple chaining of filters.
+
+Currently, filters are defined in ``lsst.alert.stream.filters`` in the
+alert_stream repository.
+Each filter is constructed as a class with a filter function.
+Below is an example of code defining a simple filter.
+
+.. code-block:: Python
+
+  class Filter001(AlertFilter):
+      def filter(self, alert):
+          if ((alert['diaSource']['snr'] > 5) &
+              (alert['diaSource']['diffFlux'] > 0.00003631)):  # 20th mag
+              return True
+          else:
+              return False
+
+Filters inherit from an AlertFilter base class.
+When the filter class is called on each alert, the visit ID is read.
+Up to 20 alerts per visit that pass the filter, i.e., return True, are
+forwarded to a topic for the filtered alerts, named for the class.
+In the above example, alerts are sent to a new topic named "Filter001."
+The filter detects the beginning of a new visit when the visit ID
+changes from one alert to the next.
+This assumes that all alerts from a visit will be received before
```

Comment:
This seems like a trivial implementation detail, right? I imagine — without looking hard at the code — that you could just store a dict of `{visit: num_alerts}` to avoid this problem.

Reply:
No.  I talked with Eric also about this.  I don't think it is trivial to put the number of alerts in a visit into each alert.  I had been thinking that one of the plusses with using Kafka is that it can act as a cache where the end of each parallel-by-ccd pipeline dumps their alert packets as they are generated.  So there would be no aggregation at the end that counts all of the alerts in that visit and plugs that number into each alert.  I think you could more easily put in the number of alerts per visit per ccd, since all the alerts from a ccd would probably be processed at the same time, but I don't think that gains you anything.  It is easy to just keep a little switch at the end that detects when the visit has changed.